components/org.wso2.carbon.identity.auth.otp.core/src/main/java/org/wso2/carbon/identity/auth/otp/core/AbstractOTPAuthenticator.java

```diff
@@ -0,0 +1,1278 @@
+                                                 HttpServletResponse response, AuthenticationContext context)
+            throws AuthenticationFailedException {
+
+        AuthenticatedUser authenticatedUserFromContext = getAuthenticatedUserFromContext(context);
+        String applicationTenantDomain = context.getTenantDomain();
+
+        /*
+         * If an invalid user has attempted to log in by submitting an OTP code, or if an invalid user has requested
+         * to resend the OTP code, they should be redirected to the OTP login page. This is only valid when OTP is
+         * set as the first step of the authentication flow because in other cases, there will always be a valid user in
+         * the context.
+         */
+        if (isLoginAttemptByInvalidUser(context, authenticatedUserFromContext)) {
+            AuthenticatedUser invalidUser = new AuthenticatedUser();
+            invalidUser.setUserName((String) context.getProperty(INVALID_USERNAME));
+            redirectToOTPLoginPage(invalidUser, applicationTenantDomain, false,
+                    response, request, context);
+            return;
+        }
+        if (this instanceof PasswordlessOTPAuthenticator && isAuthenticatorEnabledForFirstFactor()) {
+            if (authenticatedUserFromContext == null) {
+                /*
+                 * If there is no authenticated user, and the request is not returned from the Identifier First page,
+                 * redirect the user to the Identifier First page to retrieve the username.
+                 */
+                if (!isUserRedirectedFromIDF(request)) {
+                    redirectUserToIDF(response, context);
+                    context.setProperty(IS_IDF_INITIATED_FROM_AUTHENTICATOR, true);
+                    return;
+                }
+                /*
+                 * If the request is returned from the Identifier First page, resolve the user and set them in context.
+                 */
+                context.removeProperty(IS_IDF_INITIATED_FROM_AUTHENTICATOR);
+                AuthenticatedUser authenticatedUser = resolveUserFromRequest(request, context);
+                authenticatedUserFromContext = resolveUserFromUserStore(authenticatedUser);
+                setResolvedUserInContext(context, authenticatedUserFromContext);
+            } else if (isPreviousIdPAuthenticationFlowHandler(context)) {
+                /*
+                 * If the previous authentication has only been done by AuthenticationFlowHandlers, need to check if the
+                 * user exists in the database.
+                 */
+                authenticatedUserFromContext = resolveUserFromUserStore(authenticatedUserFromContext);
+                setResolvedUserInContext(context, authenticatedUserFromContext);
+            }
+            /*
+             * If the authenticated user is still null at this point, then an invalid user is trying to log in.
+             */
```

Comment:
Make it a single line comment

Reply:
Fixed with 6239874dcb1bf208bfccc91c0af83bf6b6747dbc